libs/wire-api/src/Wire/API/Conversation/Action.hs

```diff
@@ -14,74 +14,179 @@
+        ]
+
+instance ToJSON ConversationActionTag where
+  toJSON = schemaToJSON
+
+instance FromJSON ConversationActionTag where
+  parseJSON = schemaParseJSON
+
+$(genSingletons [''ConversationActionTag])
+
+-- | We use this type family instead of a sum type to be able to define
+-- individual effects per conversation action. See 'HasConversationActionEffects'.
+type family ConversationAction (tag :: ConversationActionTag) :: * where
+  ConversationAction 'ConversationJoinTag = ConversationJoin
+  ConversationAction 'ConversationLeaveTag = ConversationLeave
+  ConversationAction 'ConversationMemberUpdateTag = ConversationMemberUpdate
+  ConversationAction 'ConversationDeleteTag = ConversationDelete
+  ConversationAction 'ConversationRenameTag = ConversationRename
+  ConversationAction 'ConversationMessageTimerUpdateTag = ConversationMessageTimerUpdate
+  ConversationAction 'ConversationReceiptModeUpdateTag = ConversationReceiptModeUpdate
+  ConversationAction 'ConversationAccessDataTag = ConversationAccessData
+  ConversationAction 'ConversationRemoveMembersTag = ConversationRemoveMembers
+
+data SomeConversationAction where
+  SomeConversationAction :: Sing tag -> ConversationAction tag -> SomeConversationAction
+
+instance Show SomeConversationAction where
+  show (SomeConversationAction tag action) =
+    $(sCases ''ConversationActionTag [|tag|] [|show action|])
+
+instance Eq SomeConversationAction where
+  (SomeConversationAction tag1 action1) == (SomeConversationAction tag2 action2) =
+    case (tag1, tag2) of
+      (SConversationJoinTag, SConversationJoinTag) -> action1 == action2
+      (SConversationJoinTag, _) -> False
+      (SConversationLeaveTag, SConversationLeaveTag) -> action1 == action2
+      (SConversationLeaveTag, _) -> False
+      (SConversationMemberUpdateTag, SConversationMemberUpdateTag) -> action1 == action2
+      (SConversationMemberUpdateTag, _) -> False
+      (SConversationDeleteTag, SConversationDeleteTag) -> action1 == action2
+      (SConversationDeleteTag, _) -> False
+      (SConversationRenameTag, SConversationRenameTag) -> action1 == action2
+      (SConversationRenameTag, _) -> False
+      (SConversationMessageTimerUpdateTag, SConversationMessageTimerUpdateTag) -> action1 == action2
+      (SConversationMessageTimerUpdateTag, _) -> False
+      (SConversationReceiptModeUpdateTag, SConversationReceiptModeUpdateTag) -> action1 == action2
+      (SConversationReceiptModeUpdateTag, _) -> False
+      (SConversationAccessDataTag, SConversationAccessDataTag) -> action1 == action2
+      (SConversationAccessDataTag, _) -> False
+      (SConversationRemoveMembersTag, SConversationRemoveMembersTag) -> action1 == action2
+      (SConversationRemoveMembersTag, _) -> False
+
+instance ToJSON SomeConversationAction where
+  toJSON (SomeConversationAction sb action) =
+    let tag = fromSing sb
+        actionJSON :: Value =
+          $(sCases ''ConversationActionTag [|sb|] [|toJSON action|])
```

Comment:
The use of TH here and in similar places could be avoided. The trick is to define the schema by cases on a tag singleton, instead of defining it directly in an instance. Something like:
```haskell
conversationActionSchema :: forall tag. Sing tag → ValueSchema SwaggerDoc (ConversationAction tag)
conversationActionSchema SConversationJoin = ... definition of schema for join ...
...
```

You probably don't need the individual instances, so you end up actually with *shorter* code, if everything goes well.

Then here you could write:
```haskell
actionJSON = fromMaybe Null $ schemaOut (conversationActionSchema sb)
```
without having to pattern-match.

Let me know if you need help.